### PR TITLE
Update L1 tracking for HL-LHC (replacement of #34028)

### PR DESCRIPTION
**Pull request by skinnari**
#### Important: 

Must be coupled to, and tested with, cms-data/L1Trigger-TrackFindingTracklet#3

#### PR description:

This PR is a fixed version of #34028 (git mess-up by me ... apologies!)

This PR includes updates to the L1 tracking for HL-LHC:
- use DTC emulation (L1Trigger/TrackerDTC), the step between the Tracker and the L1 tracking system, along with associated, necessary changes to the L1 tracking code
- algorithm updates to match HLS-based version (i.e. firmware)
- code cleanup (removal of obsolete code, general code improvements)

#### PR validation:

- validated via L1 tracking performance studies on ttbar PU=200 samples.
- DQM histogram will show many "failed" as they check for any differences, but DQM checks by experts are as expected.
 
#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not a backport.

### CODE REVIEW

**Comment by srimanob** [comment]
test parameters:

- pull_request = cms-data/L1Trigger-TrackFindingTracklet#3

**Comment by srimanob** [comment]
Please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by skinnari** [comment]
are there anything further that is needed to be done for this PR before it can be merged? thanks, louise

File: L1Trigger/TrackFindingTracklet/interface/Util.h [hunk 1]
```diff
@@ -35,150 +39,142 @@ namespace trklet {
...
+        hexword = "1011";
+      else if (word == 'C')
+        hexword = "1100";
+      else if (word == 'D')
+        hexword = "1101";
+      else if (word == 'E')
+        hexword = "1110";
+      else if (word == 'F')
+        hexword = "1111";
+      else if (hexword.empty()) {
```
Comment (srimanob) [inline L1Trigger/TrackFindingTracklet/interface/Util.h:97]: Should not have else?

File: L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  pt_ = pt;
+  eta_ = eta;
+  phi_ = phi;
+  vx_ = vx;
+  vy_ = vy;
+  vz_ = vz;
+}
+
+void L1SimTrack::write(ofstream& out) {
+  if (pt_ > -2.0) {
```
Comment (srimanob) [inline L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc:26]: What does negative pt_ mean?

File: L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    out << "SimTrack: " << eventid_ << "\t" << trackid_ << "\t" << type_ << "\t" << pt_ << "\t" << eta_ << "\t" << phi_
+        << "\t" << vx_ << "\t" << vy_ << "\t" << vz_ << "\t" << endl;
+  }
+}
+
+void L1SimTrack::write(ostream& out) {
+  if (pt_ > -2) {
```
Comment (srimanob) [inline L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc:33]: Same question as above, also should it be float as above?

File: L1Trigger/TrackFindingTracklet/src/MatchEngine.cc [hunk 4]
```diff
@@ -298,40 +226,55 @@ void MatchEngine::execute() {
...
+          passz = std::abs(idrz) <= drzcut;
         } else {
-          passz = idrz >= -5 && idrz <= 5;
+          constexpr int drzcut = 3;
+          passz = std::abs(idrz) <= drzcut;
         }
       }
 
+      if (print) {
+        cout << "istep index : " << istep << " " << index << " " << vmstub.bend().value()
```
Comment (srimanob) [inline L1Trigger/TrackFindingTracklet/src/MatchEngine.cc:265]: Use Log instead?

File: L1Trigger/TrackFindingTracklet/src/MatchEngineUnit.cc [hunk 5]
```diff
@@ -69,19 +140,28 @@ void MatchEngineUnit::step() {
     }
   }
 
+  // Detailed printout for comparison with HLS code
```
Comment (srimanob) [inline L1Trigger/TrackFindingTracklet/src/MatchEngineUnit.cc:143]: If no need, please clean up. GIT will save history anyways. If it is still needed, maybe define DEBUG and use it to trigger the log.

File: L1Trigger/TrackFindingTracklet/interface/Util.h (hunk 1 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/interface/Util.h:97]: changed to else

File: L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc (hunk 2 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc:26]: it means it is taking into account the charge.

File: L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc (hunk 3 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/L1SimTrack.cc:33]: it is accounting for the charge, and it using double's for both.

File: L1Trigger/TrackFindingTracklet/src/MatchEngine.cc (hunk 4 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/MatchEngine.cc:265]: fixed

File: L1Trigger/TrackFindingTracklet/src/MatchEngineUnit.cc (hunk 5 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/MatchEngineUnit.cc:143]: changed to messagelogger

**Comment by srimanob** [comment]
Please test

**Comment by smuzaffar** [comment]
please test

**Comment by qliphy** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
+Upgrade

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
@skinnari with this PR and https://github.com/cms-data/L1Trigger-TrackFindingTracklet/pull/3 
we got many failures in IB, e.g.

https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-07-07-1100/pyRelValMatrixLogs/run/23234.103_TTbar_14TeV+2026D49Aging3000+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal/step2_TTbar_14TeV+2026D49Aging3000+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal.log#/

RuntimeError: An exception of category 'FileInPathError' occurred.
Exception Message:
edm::FileInPath unable to find file L1Trigger/TrackFindingTracklet/data/dtclinklayerdisk.dat anywhere in the search path.

It seem you deleted dtclinklayerdisk.dat  with https://github.com/cms-data/L1Trigger-TrackFindingTracklet/pull/3 
Can you have a quick look and fix? Thanks!

**Comment by qliphy** [comment]
@skinnari 

Looking again,  

CMSSW_12_0_X_2021-07-07-1100  included
https://github.com/cms-sw/cmsdist/pull/7109

while only from the next IB includes #34215
CMSSW_12_0_X_2021-07-07-1900

so we probably should wait for the new IB test to check the results.

**Comment by smuzaffar** [comment]
@qliphy , the cmssw PR was merged after 11h00 CERN time that is why it was not picked up for 11h00 IB. In future when you need to merge changes which requires externals updates then please do 10-15 mins before the hour e.g. etither at 10h45 ( for 11h00 IB) or 22h45 (for 23h00 IB).

**Comment by skinnari** [comment]
hi, just wanted to confirm that everything was indeed fine with the new IBs? please let me know otherwise, thanks!

**Pull request merged**